erpnext/public/js/utils/manufacturing.js

```diff
@@ -470,5 +676,68 @@ $.extend(erpnext.manufacturing, {
 
 	has_stock_entry_permission: function () {
 		return frappe.model.can_write("Stock Entry");
-	}
+	},
+	finish_operation: function(doc) {
+		if (doc.docstatus != 1) {
+			return;
+		}
+		// frappe.prompt(fields, data => {
+		// 	resolve();
+		// }, __('Finish Operation'), __('Finish'));
+	},
 });
+
+
+function show_transition_dialog(doc, purpose) {
+	var dialog = new frappe.ui.Dialog({
+		title: __('Finish What?'),
+		fields: [
+			{
+				label: __('Select Action'),
+				fieldname: 'action',
+				fieldtype: 'Select',
+				options: '\nFinish Operation\nFinish Work Order',
+			},
+		],
+		primary_action: function() {
+			let d = dialog.get_values();
+			if (d.action == "Finish Operation") {
+				return show_prompt_for_work_order_operation(doc);
+			} else if (d.action == "Finish Work Order") {
+				return show_prompt_for_work_order_stock_entry(doc, purpose);
+			}
+			dialog.hide()
+		},
+		primary_action_label: __('Finish')
+	});
+	dialog.show();
+}
+
+function show_prompt_for_work_order_stock_entry(doc, purpose) {
+	return erpnext.manufacturing.show_qty_prompt_for_stock_entry(doc, purpose).then(r => {
+		return frappe.call({
+			method: "erpnext.manufacturing.doctype.work_order.work_order.make_stock_entry",
+			args: {
+				"work_order_id": doc.name,
+				"purpose": purpose,
+				"scrap_remaining": r.data.scrap_remaining,
+				"qty": r.data.qty,
+				"args": r.args,
+			},
+			freeze: 1,
+			callback: (r) => {
+				if (r.message) {
+					frappe.model.sync(r.message);
+
+					if (cur_frm && cur_frm.doc.doctype == "Work Order" && cur_frm.doc.name == doc.name) {
+						cur_frm.reload_doc();
+					}
+
+					if (r.message.docstatus != 1) {
+						frappe.set_route('Form', r.message.doctype, r.message.name);
+					}
+				}
+			}
+		});
+	});
+}
```

Comment:
why global scoped function .. why not use `erpnext.manufacturing` namespace? temporary functions?